presto-main/src/main/java/com/facebook/presto/sql/planner/NodePartitioningManager.java

```diff
@@ -258,9 +258,25 @@ private ConnectorBucketNodeMap getConnectorBucketNodeMap(Session session, Partit
         return distribution.build();
     }
 
-    public List<Node> sortedNodes(Session session, ConnectorId connectorId)
+    public List<Node> getNodes(Session session, ConnectorId connectorId)
     {
-        List<InternalNode> nodes = nodeScheduler.createNodeSelector(session, connectorId).allNodes();
-        return nodes.stream().sorted(comparing(InternalNode::getNodeIdentifier)).collect(toImmutableList());
+        // Nodes returned by the node selector are already sorted based on nodeIdentifier. No need to sort again
+        List<InternalNode> allNodes = nodeScheduler.createNodeSelector(session, connectorId).getAllNodes();
+
+        ImmutableList.Builder<Node> nodeBuilder = ImmutableList.builder();
+        int nodeCount = allNodes.size();
+        for (int i = 0; i < nodeCount; i++) {
+            InternalNode node = allNodes.get(i);
+            if (node.getNodeStatus() == DEAD) {
+                // Replace dead nodes with the first alive node to the right of it in the sorted node list
+                int index = (i + 1) % nodeCount;
+                while (node.getNodeStatus() == DEAD && index < nodeCount) {
```

Comment:
I am not sure if this is the correct logic. With last node being dead, this will resume the loop again. In this case it would add node 0 two times. Why don't we just filter out dead nodes? Why do we need to maintain index?

Reply:
We want the size of the nodes list to be same constant even if a node goes down right. That is the basic premise of the PR right. That is the reason we dont remove the dead nodes.

Reply:
You are not really maintaining indexes here. Here we are just removing dead nodes, it seems. And that flow as well has a bug of counting nodes multiple times.

Reply:
I had a offline discussion with Nikhil on this. We will move this logic from here to scheduler. We will also try to keep this logic similar to current affinity scheduling.

Reply:
Discussed offline. It is very hard to do dead node handling at per split level for a bucketed table. Moving this logic to NodeSelector may not be correct because split scheduling is already in progress by that time. Will continue with this current change and may be will comeback to it in some later PR.